ChronoAPI/ChronoLog/include/ClocksourceManager.h

```diff
@@ -10,94 +10,117 @@
+	{
+	    auto t = std::chrono::high_resolution_clock::now().time_since_epoch().count();
+	    return t;
+	}
 };
 
-class ClocksourceTSC : public Clocksource {
-public:
-    uint64_t getTimestamp() override {
-        unsigned int proc_id;
-        uint64_t  t = __builtin_ia32_rdtscp(&proc_id);
-        lfence();
-        return t;
-    }
-};
-class ClocksourceCPPStyle : public Clocksource {
+template<class ClockSource>
+class ClocksourceManager {
 public:
-    uint64_t getTimestamp() override {
-        using namespace std::chrono;
-        using clock = steady_clock;
-        clock::time_point t = clock::now();
-        return (t.time_since_epoch().count());
-    }
-};
+	ClocksourceManager(std::string &unit_name)
+	{
+	  unit = 1;
+	  if(unit_name.compare("nanoseconds")==0) unit = 1;
+	  else if(unit_name.compare("microseconds")==0) unit = 1000;
+	  else if(unit_name.compare("milliseconds")==0) unit = 1000000;
+	  else if(unit_name.compare("seconds")==0) unit = 1000000000;
+	  offset = 0;
+	  pos_neg = false;
+	  func_prefix = "ChronoLogThallium";
+	}
+	~ClocksourceManager()
+	{}
+	uint64_t TimeStamp()
+	{
+		if(!pos_neg)
+		return clocksource_->getTimeStamp()/unit+offset;
+		else return clocksource_->getTimeStamp()/unit-offset;
+	}
+	uint64_t LocalGetTS(std::string &client_id)
+	{
+	     uint64_t t = TimeStamp();
 
-class ClocksourceManager {
-private:
-    ClocksourceManager() : clocksource_(nullptr), clocksourceType_(ClocksourceType::C_STYLE) {
-        LOGD("%s constructor is called", typeid(*this).name());
-    }
+	     return t;
 
-public:
-    ~ClocksourceManager() {
-        if (clocksource_) {
-            delete clocksource_;
-            clocksource_ = nullptr;
-        }
-        if (clocksourceManager_) {
-            delete clocksourceManager_;
-            clocksourceManager_ = nullptr;
-        }
-    }
+	}
+
+        void bind_functions(std::shared_ptr<ChronoLogRPC> rpc)
+	{
+		std::function<void(const tl::request &,
+				   std::string &
+                                   )> timestampFunc(
+                        [this](auto && PH1,
+			       auto && PH2
+				) {
+                            ThalliumLocalGetTS(std::forward<decltype(PH1)>(PH1),
+					    std::forward<decltype(PH2)>(PH2));
+			    });
 
-    static ClocksourceManager *getInstance() {
-        if (!clocksourceManager_) {
-            clocksourceManager_ = new ClocksourceManager();
-        }
-        return clocksourceManager_;
-    }
+		rpc->bind("ChronoLogThalliumGetTS", timestampFunc);
+	}
 
-    void setClocksourceType(ClocksourceType type) {
-        this->clocksourceType_ = type;
-    }
 
-    ClocksourceType getClocksourceType() {
-        return clocksourceType_;
-    }
+	CHRONOLOG_THALLIUM_DEFINE(LocalGetTS,(client_id),std::string& client_id)
 
-    Clocksource *getClocksource() {
-        this->clocksource_ = Clocksource::Create(this->clocksourceType_);
-        return this->clocksource_;
-    }
+	void set_offset(uint64_t &o, bool &b)
+	{
```

Comment:
can offset be int64_t instead of uint64_t so that it can hold the sign bit  instead of introducing pos_neg variable?

Reply:
Could you clarify who owns/creates/deletes Clocksource ?

Reply:
The ClockManager is  instantiated in the constructors of visor and client classes. It is deleted in the destructors.

Reply:
I think both Clocksource and ClocksourceManager should be instantiated before any of the RPC objects are .

Reply:
> can offset be int64_t instead of uint64_t so that it can hold the sign bit instead of introducing pos_neg variable?

I guess @aparnasasidharan2017 wants to maximize the range of offset by manually "adding" sign to uint64_t. I agree that it can do that. But I don't think it is necessary. It is impossible to have an offset more than half of the max timestamp. ChronoLog should prevent it from happening way earlier. I think int64_t is good enough.

Reply:
@Kun timestamps are never negative. They should be unsigned numbers ranging from 0 to the largest value of the datatype (64 bits here).

Reply:
@aparnasasidharan2017 I understand that timestamps will be never be negative and the differences will. My point was having a bool sign and a uint64_t difference does not bring us more power in handling our case in ChronoLog than a int64_t. The 50% reduction in range does not hurt us because it will be extremely unlikely to have one clock at MAX_UINT64 and another at 0. And our system should prevent such a terrible case from happening at the beginning.